torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
                     else:
                         p_assert(
-                            not p._is_sharded, "All sharded parameters should "
+                            not p._is_sharded or not getattr(p, '_post_backward_called', False), "All sharded parameters that received gradient should "
                             "use `_saved_grad_shard`"
                         )
                     if hasattr(p, "_saved_grad_shard"):
                         delattr(p, "_saved_grad_shard")
+                    if hasattr(p, '_post_backward_called'):
+                        delattr(p, '_post_backward_called')
 
         # Update root and nested FSDP's hooks and flags.
         for m in self.modules():  # includes self
             if isinstance(m, FullyShardedDataParallel):
-                _finalize_params(m)
-                m._pre_backward_hook_has_run = False
                 if any(p.requires_grad for p in m.parameters()):
                     # Check if the module has params and if any of them has
                     # the `requires_grad` field set. If `requires_grad=False` for
                     # all the params, the post_backward hook will not fire and the
                     # state will remain in `TrainingState_.BACKWARD_PRE`.
-                    if any([p.requires_grad for p in m.params]):
-                        m._assert_state(TrainingState_.BACKWARD_POST)
+                    managed_param_requires_grad = any(p.requires_grad for p in m.params)
+                    if managed_param_requires_grad:
+                        hook_called = any(
+                            getattr(p, '_post_backward_called', False) for p in m.params
+                        )
+                        if hook_called:
+                            m._assert_state(TrainingState_.BACKWARD_POST)
+                            # Delete attr so the logic works correctly next iteration
+                            for p in m.params:
+                                if hasattr(p, '_post_backward_called'):
+                                    delattr(p, '_post_backward_called')
+                        else:
+                            # post backward hook was not called, meaning param
```

Comment:
nit: The comment's grammar assumes single parameter per `FullyShardedDataParallel`, while the code still iterates over `.params`. I prefer that the comment assumes plural parameters per `FullyShardeDataParallel` to match the code and because this will be true for non-recursive wrapping.